### PR TITLE
Bow OpenAPI - Tests for auto-generated code

### PR DESCRIPTION
**Pull request by miguelangel-dev**
## Testing
Added tests for:
- API Configuration
- API Helpers
- API Client: send operation

## Library for testing
We have added a group of files to help to client in the testing. They will be generated into `XCTest` folder.
- `HTTPErrorTest`
- `asserts` for comparing API result
- `stub` helpers for mocking URLSession

### CODE REVIEW

**Comment by miguelangel-dev** [comment]
@truizlop could you do a final review to this PR?

File: Bow OpenAPI/Generator/Algebra/ClientGenerator.swift [hunk 1]
```diff
@@ -4,6 +4,11 @@ import Foundation
 import Bow
 import BowEffects
 
+struct GeneratorOutput {
```
Comment (truizlop) [inline Bow OpenAPI/Generator/Algebra/ClientGenerator.swift]: Perhaps name it `OutputPath`?

File: Bow OpenAPI/Generator/Algebra/FileSystem.swift [hunk 2]
```diff
@@ -37,4 +38,11 @@ extension FileSystem {
     func removeFiles(_ files: String...) -> IO<FileSystemError, ()> {
         files.traverse(remove(itemPath:)).void()^
     }
+    
+    func rename(_ name: String, itemAt: String) -> IO<FileSystemError, ()> {
+        let copyItem = copy(itemPath: itemAt, toPath: "\(itemAt.parentPath)/\(name).\(itemAt.extension)")
+        let removeCopiedItem = removeFiles(itemAt)
+        
+        return copyItem.followedBy(removeCopiedItem)^
+    }
```
Comment (truizlop) [inline Bow OpenAPI/Generator/Algebra/FileSystem.swift]: This implementation is being too clever about renaming and assuming you want to keep the same extension, which may not always be the case. Force the user to pass the name **with** the extension. It can even be defined in terms of a `move` function:

```swift
func move(from origin: String, to destination: String) -> IO<FileSystemError, Void> {
  copy(itemPath: origin, toPath: destination)
    .followedBy(removeFiles(origin))
}

func rename(_ newName: String, itemAt: String) -> IO<FileSystemError, Void> {
  move(from: itemAt, to: "\(itemAt.parentPath)/\(newName)")
}
```

File: Bow OpenAPI/Generator/SwaggerClientGenerator.swift [hunk 3]
```diff
@@ -31,11 +31,13 @@ class SwaggerClientGenerator: ClientGenerator {
...
+    private func reorganizeFiles(in outputPath: GeneratorOutput, fromTemplate templatePath: String) -> EnvIO<FileSystem, APIClientError, ()> {
         EnvIO { fileSystem in
             binding(
-                |<-fileSystem.move(from: "\(output)/SwaggerClient/Classes/Swaggers", to: output),
-                |<-fileSystem.remove(from: output, files: "Cartfile", "AlamofireImplementations.swift", "Models.swift", "git_push.sh", "SwaggerClient.podspec", "SwaggerClient", ".swagger-codegen", ".swagger-codegen-ignore", "JSONEncodableEncoding.swift", "JSONEncodingHelper.swift"),
+                |<-fileSystem.move(from: "\(outputPath.sources)/SwaggerClient/Classes/Swaggers", to: outputPath.sources),
+                |<-fileSystem.remove(from: outputPath.sources, files: "Cartfile", "AlamofireImplementations.swift", "Models.swift", "git_push.sh", "SwaggerClient.podspec", "SwaggerClient", ".swagger-codegen", ".swagger-codegen-ignore", "JSONEncodableEncoding.swift", "JSONEncodingHelper.swift"),
+                |<-fileSystem.rename("APIConfiguration", itemAt: "\(outputPath.sources)/APIHelper.swift"),
```
Comment (truizlop) [inline Bow OpenAPI/Generator/SwaggerClientGenerator.swift]: According to my comment above, this file name should include the `.swift` extension.

File: Templates/XCTest/API+XCTest.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                              succeeds: T,
+                              _ information: String? = nil, file: StaticString = #file, line: UInt = #line) {
+        envIO.provide(apiConfig)
+             .unsafeRunSyncEither()
+             .assert(success: succeeds, information: information, file: file, line: line)
+    }
+    
+    func assert<T: Equatable>(_ envIO: EnvIO<API.Config, API.HTTPError, T>,
+                              withConfig apiConfig: API.Config,
+                              failures: HTTPErrorTest,
```
Comment (truizlop) [inline Templates/XCTest/API+XCTest.swift]: Replace `failures` with `fails`.

File: Tests/Utils/API+Generators.swift [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    }
+}
+
+
+// MARK: Common <generator>
+extension Data: Arbitrary {
+    public static var arbitrary: Gen<Data> {
+        let data = String.arbitrary.generate.data(using: .utf8)!
+        return Gen<Data>.pure(data)
+    }
```
Comment (truizlop) [inline Tests/Utils/API+Generators.swift]: A better approach:
```swift
String.arbitrary.map { str in str.data(using: .utf8) }
```

File: Tests/Utils/APITestCase.swift [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    
+    // MARK: send operation
+    func send<T: Codable>(request: URLRequest) -> Either<API.HTTPError, T> {
+        let envIO = EnvIO<API.Config, API.HTTPError, T> { config in
+            API.send(request: request, session: config.session, decoder: config.decoder)
+        }
+            
+        return envIO.provide(self.apiConfig).unsafeRunSyncEither()
+    }
+}
```
Comment (truizlop) [inline Tests/Utils/APITestCase.swift]: Is this file needed anymore?

File: Templates/XCTest/APIConfigTesting.swift [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        StubURL.stub(contentsOfFile: url, code: code)
+        return copy(session: URLSession(configuration: URLSessionConfiguration.testing))
+    }
+}
+
+public extension API.Config {
+    var request: URLRequest {
+        return URLRequest(url: URL(string: basePath)!)
+    }
+}
```
Comment (truizlop) [inline Templates/XCTest/APIConfigTesting.swift]: I wouldn't put this as an extension field of `API.Config` as this is a fixture of a particular test.

File: Templates/XCTest/APIConfigTesting.swift [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+import Foundation
+
+public extension API.Config {
+    
+    /// Stubs data to respond the next HTTP request with.
+    /// - Parameter data: Contents of the response to the next request.
+    /// - Parameter code: HTTP status code for the response. Defaults to 200.
+    func stub(data: Data, code: Int = 200) -> API.Config {
+        StubURL.reset()
```
Comment (truizlop) [inline Templates/XCTest/APIConfigTesting.swift:11]: This is nice at the moment and for this PR, but we could consider a more general approach where you can stub multiple results to test an orchestration of different calls and stubbing a particular endpoint. Something like:

```swift
stub("/user/profile", data: Data(), code: 200)
stub("/user/images", data: Data(), code: 200)
```

What do you think?

**Review by truizlop — CHANGES_REQUESTED** [review]

File: Templates/XCTest/API+XCTest.swift (hunk 4 above)
Comment (miguelangel-dev) [reply Templates/XCTest/API+XCTest.swift]: I thought I have changed it -- ups!

File: Templates/XCTest/APIConfigTesting.swift (hunk 8 above)
Comment (miguelangel-dev) [reply Templates/XCTest/APIConfigTesting.swift:11]: Yep, it sounds good; we could create a queue where adding the associate endpoint-response

File: Tests/Utils/APITestCase.swift (hunk 6 above)
Comment (miguelangel-dev) [reply Tests/Utils/APITestCase.swift]: In fact, no, I removed the reference, but I do not delete to the project. Good catch!

File: Templates/XCTest/APIConfigTesting.swift (hunk 7 above)
Comment (miguelangel-dev) [reply Templates/XCTest/APIConfigTesting.swift]: I agree

**Review by truizlop — APPROVED** [review]
Awesome!
